esp/esp/program/modules/handlers/studentregphasezeromanage.py

```diff
@@ -58,32 +58,56 @@ def module_properties(cls):
-        # If any one in the group doesn't get in (due to cap size), no one in that group gets in.
-        counts = {key:0 for key in grade_caps}
+    def lottery(self, prog, role, post):
+        messages = []
         winners, _ = Group.objects.get_or_create(name=role)
 
-        for i in records:
-            sibs = i.user.all()
-            newcounts = copy.copy(counts)
-            for j in sibs:
-                newcounts[j.getGrade(prog)] += 1
-
-            cpass = not any(newcounts[c] > grade_caps[c] for c in counts)
-
-            if cpass:
-                winners.user_set.add(*sibs)
-                counts = newcounts
+        if post.get('mode') == 'default':
+            # Get grade caps
+            grade_caps_str = prog.grade_caps()
+            if grade_caps_str:
+                grade_caps = {int(key[0]): value for key, value in grade_caps_str.iteritems()}
+                ###############################################################################
+                # The default lottery algorithm is run, with randomization and processing in order.
+                # If any one in the group doesn't get in (due to cap size), no one in that group gets in.
+                records = PhaseZeroRecord.objects.filter(program=prog).order_by('?')
+
+                counts = {key:0 for key in grade_caps}
+
+                for i in records:
+                    sibs = i.user.all()
+                    newcounts = copy.copy(counts)
+                    for j in sibs:
+                        newcounts[j.getGrade(prog)] += 1
+
+                    cpass = not any(newcounts[c] > grade_caps[c] for c in counts)
+
+                    if cpass:
+                        winners.user_set.add(*sibs)
+                        counts = newcounts
+            else:
+                messages.append("<i>program_size_by_grade</i> <a href='/manage/" + prog.getUrlBase() + "/tags'>tag</a> is not set. Lottery not run.")
+                return messages
+
+        elif post.get('mode') == 'manual':
+            usernames = filter(None, re.split(r'[;,\s]\s*', post.get('usernames')))
+
+            for username in usernames:
+                try:
+                    student = ESPUser.objects.get(username=username)
+                except (ValueError, ESPUser.DoesNotExist):
+                    messages.append("Could not find a user with username " + username)
+                    continue
+                if student.isStudent():
+                    if post.get('groups'):
+                        records = PhaseZeroRecord.objects.filter(program=prog, user=student).order_by('time')
+                        if records.count() > 0:
+                            group_members = records[0].user.all()
+                            winners.user_set.add(*group_members)
+                    else:
+                        winners.user_set.add(student)
+                else:
+                    messages.append(username + "is not a student")
```

Comment:
Do you think it's a better idea to validate all users first and reject without making any changes at all if any of them are invalid (user doesn't exist, not a student, etc.) ?  An invalid user points to an error in the manual lottery algorithm, and since it's difficult to undo the lottery process, it might be better to just reject completely.

Reply:
That's a good idea. I'll change the code around to check the students first, then add to the winners group only if all checks pass.